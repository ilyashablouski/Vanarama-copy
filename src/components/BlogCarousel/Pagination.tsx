import React from 'react';
import Icon from 'core/atoms/icon';
import ChevronBack from 'core/assets/icons/ChevronBack';
import ChevronForward from 'core/assets/icons/ChevronForward';
import { IBaseProps } from 'core/interfaces/base';

const Pagination = ({ dataUiTestId }: IBaseProps) => {
  return (
    <div className="carousel--nav">
      <button
        className="carousel--nav-arrow swiper-prev"
        type="button"
        data-uitestid={
          dataUiTestId
            ? `${dataUiTestId}_swiper-pagination_arrow-prev`
            : undefined
        }
      >
        <Icon icon={<ChevronBack />} color="darker" size="regular" />
      </button>

      <div
        className="swiper-pagination"
        data-uitestid={
          dataUiTestId ? `${dataUiTestId}_swiper-pagination` : undefined
        }
      />

      <button
        className="carousel--nav-arrow swiper-next"
        type="button"
        data-uitestid={
          dataUiTestId
            ? `${dataUiTestId}_swiper-pagination_arrow-next`
            : undefined
        }
      >
        <Icon icon={<ChevronForward />} color="darker" size="regular" />
      </button>
    </div>
  );
};

export default Pagination;
