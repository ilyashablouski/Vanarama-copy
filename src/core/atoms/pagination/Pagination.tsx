import React, { FC, useCallback, useMemo } from 'react';
import cx from 'classnames';
import { IPaginationProps } from './interfaces';
import Link from '../link';
import ChevronBackSharp from '../../assets/icons/ChevronBackSharp';
import ChevronForwardSharp from '../../assets/icons/ChevronForwardSharp';
import Icon from '../icon';

const Pagination: FC<IPaginationProps> = props => {
  const {
    className,
    pages,
    selected,
    path,
    pathWithHtml,
    pathForFirstPage,
    onClick,
    onClickBackArray,
    onClickNextArray,
  } = props;

  const addEtc = (pagesArray: any[]) => {
    if (pagesArray.length <= 1) {
      return pagesArray;
    }
    const lastPage = pagesArray[pagesArray.length - 1];
    const secondLast = pagesArray[pagesArray.length - 2];
    if (pagesArray[0] + 1 !== pagesArray[1]) {
      pagesArray.splice(1, 0, '...');
    }
    if (lastPage !== secondLast + 1) {
      pagesArray.splice(pagesArray.length - 1, 0, '...');
    }
    return pagesArray;
  };

  const genPages = () => {
    const totalPage = pages.length;
    const first = (page: number) => page === 1;
    const middle = (page: number, between: { first: number; last: number }) =>
      page > between.first && page <= between.last;
    const last = (page: number) => page === totalPage;
    const between = {
      first: 0,
      last: 0,
    };
    if (totalPage <= 5) {
      between.first = 1;
      between.last = 5;
    } else if (selected < 3) {
      between.first = 1;
      between.last = 4;
    } else if (selected > totalPage - 3) {
      between.first = totalPage - 4;
      between.last = totalPage - 1;
    } else {
      between.first = selected - 2;
      between.last = selected + 1;
    }
    const allPages = pages.filter(
      page => first(page) || middle(page, between) || last(page),
    );
    return addEtc(allPages);
  };

  const genPath = useCallback(
    (pageNumber: number) =>
      `${path}/${pageNumber}${pathWithHtml ? '.html' : ''}`,
    [path, pathWithHtml],
  );

  const pagesItem = genPages();

  const previousPagePath = useMemo(() => {
    if (selected === 1) {
      return path;
    }
    if (selected === 2 && pathForFirstPage) {
      return `${pathForFirstPage}${pathWithHtml ? '.html' : ''}`;
    }
    return genPath(selected - 1);
  }, [selected, genPath, path, pathForFirstPage, pathWithHtml]);

  const nextPagePath = useMemo(() => {
    if (selected === pages[pages.length - 1]) {
      return path;
    }
    return genPath(selected + 1);
  }, [selected, genPath, pages, path]);

  return (
    <div className={cx('pagination', className)}>
      {selected !== 1 && (
        <Link
          key="chevron-back"
          href={previousPagePath}
          onClick={onClickBackArray}
          className={cx('pagination--chevron', {
            '-disabled': selected === 1,
          })}
        >
          <Icon icon={<ChevronBackSharp />} />
        </Link>
      )}
      {pagesItem.map((page: any, inx: number) => {
        return page === '...' ? (
          <div className="pagination--item --disabled" key={`${page + inx}`}>
            {page}
          </div>
        ) : (
          <Link
            key={`${page + inx}`}
            href={
              page === 1 && pathForFirstPage
                ? `${pathForFirstPage}${pathWithHtml ? '.html' : ''}`
                : `${path}/${page}${pathWithHtml ? '.html' : ''}`
            }
            onClick={onClick}
            className={cx('pagination--item', {
              '-active': selected === page,
            })}
          >
            {page}
          </Link>
        );
      })}
      {selected !== pages[pages.length - 1] && (
        <Link
          key="chevron-forward"
          href={nextPagePath}
          onClick={onClickNextArray}
          className={cx('pagination--chevron', {
            '-disabled': selected === pages[pages.length - 1],
          })}
        >
          <Icon icon={<ChevronForwardSharp />} />
        </Link>
      )}
    </div>
  );
};

export default React.memo(Pagination);
