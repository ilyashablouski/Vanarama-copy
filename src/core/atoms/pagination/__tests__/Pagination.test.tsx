import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import Pagination from '..';

const mandatoryProps = {
  pages: [1, 2, 3, 4, 5, 6, 7, 8],
  selected: 3,
  path: 'page',
};

function getComponent() {
  return renderer.create(<Pagination {...mandatoryProps} />).toJSON();
}

describe('<Pagination />', () => {
  it('renders correctly', () => {
    const tree = getComponent();
    expect(tree).toMatchSnapshot();
  });

  it('should be able to click the pagination', () => {
    const onClick = jest.fn();
    const element = mount(<Pagination {...mandatoryProps} onClick={onClick} />);
    element.find('.pagination').simulate('click');
  });
});
