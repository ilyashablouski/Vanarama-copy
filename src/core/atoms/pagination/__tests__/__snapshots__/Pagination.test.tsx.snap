// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> renders correctly 1`] = `
<div
  className="pagination"
>
  <a
    className="link pagination--chevron -primary -regular -plain"
    href="page/2"
  >
    <i
      className="icon"
    >
      <svg
        height="1em"
        viewBox="0 0 512 512"
        width="1em"
      >
        <polyline
          points="328 112 184 256 328 400"
          style={
            Object {
              "fill": "none",
              "stroke": "var(--class-color)",
              "strokeLinecap": "square",
              "strokeMiterlimit": 10,
              "strokeWidth": 48,
            }
          }
        />
      </svg>
    </i>
  </a>
  <a
    className="link pagination--item -primary -regular -plain"
    href="page/1"
  >
    1
  </a>
  <a
    className="link pagination--item -primary -regular -plain"
    href="page/2"
  >
    2
  </a>
  <a
    className="link pagination--item -active -primary -regular -plain"
    href="page/3"
  >
    3
  </a>
  <a
    className="link pagination--item -primary -regular -plain"
    href="page/4"
  >
    4
  </a>
  <div
    className="pagination--item --disabled"
  >
    ...
  </div>
  <a
    className="link pagination--item -primary -regular -plain"
    href="page/8"
  >
    8
  </a>
  <a
    className="link pagination--chevron -primary -regular -plain"
    href="page/4"
  >
    <i
      className="icon"
    >
      <svg
        height="1em"
        viewBox="0 0 512 512"
        width="1em"
      >
        <polyline
          points="184 112 328 256 184 400"
          style={
            Object {
              "fill": "none",
              "stroke": "var(--class-color)",
              "strokeLinecap": "square",
              "strokeMiterlimit": 10,
              "strokeWidth": 48,
            }
          }
        />
      </svg>
    </i>
  </a>
</div>
`;
